policy: Keep NameManager locked during SelectorCache operations

jrajahalme opened the pull request:
UpdateGenerateDNS updated the cache and generate selector updates as two
separate lock-holding sections. When updates happened concurrently, via
DNS lookups, DNS GC, or endpoint restores, the snapshot of data that
eventually updated the selectors could be out-of-order. This created an
A-B-A race for selectors that needed an update and those that needed to
be cleared.

These races have been observed on agent starts with pending pods but may
manifest between multiple pods making DNS updates or the DNS garbage
collector controller and any pod.

The code now holds the NameManager lock for the entirety of
UpdateGenerateDNS and ForceRegenerateDNS. This forces the update to
complete in its entirety before allowing another to modify the FQDN
cache or registered selectors.

The locking order between the NameManager lock and the SelectorCache
lock has been accidental so far. When removing a selector user from
the cache, the SelectorCache lock is taken, and if an FQDN selector
has no more users, the NameManager is notified of this, which
internally takes the NameManager lock. Reverse this by explicitly
locking NamaManager for the duration of selector cache operations.

From now on, if the two locks are to be held at the same time, the
NameManager mutex MUST be taken first, the SelectorCache mutex second,
as the name manager is a higher level construct than the selector
cache.

Exposing 'Lock()' and 'Unlock()' operations in an interface is not
ideal, but it helps make the lockinng order explicit, so it also
serves as documentation on the required locking order.

NameManager now return IDs also if the selector has already been
registered. This should help not get into a state where FQDNs are not
passed to the selector cache in case of a race condition.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/10501)
<!-- Reviewable:end -->

joestringer reviewed (commented):
Who owns the `fqdnSelector` here? Is it guaranteed to be retired if the `len(f.users)` goes to zero? Or can another goroutine find it and add a user while the lock is no longer held and the `selectorManager` is calling the callback?

Also, I think it would be helpful to future readers if the expectations around the `cb` were more clearly defined in comments somewhere here. It appears to be a `Finally` type function, similar in some ways to some of the regeneration code, where if this CB is non-null then the callers must call it when they're finished executing their own logic.

raybejjani commented on pkg/policy/selectorcache.go:
```diff
@@ -705,21 +704,29 @@ func (sc *SelectorCache) AddIdentitySelector(user CachedSelectionUser, selector
 func (sc *SelectorCache) RemoveSelector(selector CachedSelector, user CachedSelectionUser) {
 	sc.mutex.Lock()
-	sc.removeSelectorLocked(selector, user)
+	cb := sc.removeSelectorLocked(selector, user)
```
Edit: nevermind! I'm a dumbo and missed the change to that function

This needs to be wrapped in a `func()` doesn't it? Otherwise cb is the return of `removeSelectorLocked` but that returns nothing! (it's the none return? void?)

jrajahalme committed:
policy: Reverse locking order between selector cache and name manager

UpdateGenerateDNS updated the cache and generate selector updates as two
separate lock-holding sections. When updates happened concurrently, via
DNS lookups, DNS GC, or endpoint restores, the snapshot of data that
eventually updated the selectors could be out-of-order. This created an
A-B-A race for selectors that needed an update and those that needed to
be cleared.

These races have been observed on agent starts with pending pods but may
manifest between multiple pods making DNS updates or the DNS garbage
collector controller and any pod.

The code now holds the NameManager lock for the entirety of
UpdateGenerateDNS and ForceRegenerateDNS. This forces the update to
complete in its entirety before allowing another to modify the FQDN
cache or registered selectors.

The locking order between the NameManager lock and the SelectorCache
lock has been accidental so far. When removing a selector user from
the cache, the SelectorCache lock is taken, and if an FQDN selector
has no more users, the NameManager is notified of this, which
internally takes the NameManager lock. Reverse this by explicitly
locking NamaManager for the duration of selector cache operations.

From now on, if the two locks are to be held at the same time, the
NameManager mutex MUST be taken first, the SelectorCache mutex second,
as the name manager is a higher level construct than the selector
cache.

Exposing 'Lock()' and 'Unlock()' operations in an interface is not
ideal, but it helps make the lockinng order explicit, so it also
serves as documentation on the required locking order.

NameManager now return IDs also if the selector has already been
registered. This should help not get into a state where FQDNs are not
passed to the selector cache in case of a race condition.

Signed-off-by: Ray Bejjani <ray@isovalent.com>
Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

raybejjani commented on pkg/fqdn/name_manager.go:
```diff
@@ -95,41 +107,35 @@ func (n *NameManager) GetModel() *models.NameManager {
 // Selector will be returned as CIDR identities, as other DNS Names which have
 // already been resolved may match this FQDNSelector.
 func (n *NameManager) RegisterForIdentityUpdates(selector api.FQDNSelector) []identity.NumericIdentity {
```
Could you rename this to `RegisterForIdentityUpdatesLocked`  to make it more obvious that it doesn't lock for the caller (also Unregister).

raybejjani reviewed (approved):
Just a small nit, nothing blocking from me.

joestringer commented on pkg/policy/selectorcache_test.go:
```diff
@@ -510,17 +524,29 @@ func newDummyIdentityNotifier() *dummyIdentityNotifier {
+}
+
+// RegisterForIdentityUpdatesLocked starts managing this selector.
+func (d *dummyIdentityNotifier) RegisterForIdentityUpdatesLocked(selector api.FQDNSelector) (identities []identity.NumericIdentity) {
```
nit for future follow-up: Seems like this and `pkg/k8s/network_policy_test.go ` have dummyIdentityNotifier implementations, we could consider refactoring them into `pkg/testutils` to deduplicate.

jrajahalme replied:
Done.

joestringer commented on pkg/fqdn/name_manager.go:
```diff
@@ -87,49 +87,55 @@ func (n *NameManager) GetModel() *models.NameManager {
-		n.Mutex.Unlock()
-		return nil
-	}
+		log.WithField("fqdnSelector", selector).Warning("FQDNSelector was already registered for updates")
```
Is there any reference-counting implications for the change here for continuing for existing selectors?

Specifically for multiple calls to `ipcache.AllocateCIDRsForIPs()`.

jrajahalme replied:
Good catch, did not remember about refcounting in the allocator, this change reversed.

joestringer reviewed (commented):
One specific concern below around refcounting and CIDR identities, did you check on this before?

joestringer reviewed (approved):
LGTM
